Update ARM and AARCH64 AInvs for explicit FPU changes

corlewis opened the pull request:
This updates the AInvs proofs for ARM and AARCH64 following the specification changes made in https://github.com/seL4/l4v/pull/819 that made FPU state more explicit and changed the kernel to use semi-lazy FPU switching. It additionally proves a new invariant about the FPU that we expect to be necessary for later stages of the functional correctness proofs and also updates Access for ARM.

lsf37 commented on spec/abstract/AARCH64/Arch_Structs_A.thy:
```diff
@@ -369,13 +369,14 @@ text \<open> Arch-specific part of a TCB: this must have at least a field for us
 record arch_tcb =
   tcb_context :: user_context
   tcb_vcpu    :: "obj_ref option"
+  tcb_cur_fpu :: bool
```
We should probably have a comment in the spec what this field is used for and that it is a ghost variable (we only write to it, never read from it).

lsf37 replied:
(The comment could be mostly copy/paste from the commit message)

lsf37 commented on proof/invariant-abstract/AARCH64/ArchTcbAcc_AI.thy:
```diff
@@ -175,4 +169,110 @@ global_interpretation TcbAcc_AI?: TcbAcc_AI
   case 1 show ?case by (unfold_locales; (fact TcbAcc_AI_assms)?)
   qed
 
+context Arch begin arch_global_naming
+
+\<comment> \<open>FIXME: do we want a generic version with valid_arch_idle in the precondition\<close>
+lemma arch_thread_set_valid_idle[wp]:
+  "\<lbrace>valid_idle and (\<lambda>s. t \<noteq> idle_thread s \<or> (\<forall>atcb. tcb_vcpu atcb = None \<longrightarrow> tcb_vcpu (f atcb) = None))\<rbrace>
+   arch_thread_set f t
+   \<lbrace>\<lambda>rv. valid_idle\<rbrace>"
```
After completing the proof, do you think it would be useful? I guess we'll know more about the arch interface when we have made the other architectures work again.

corlewis replied:
I've pulled this commit into a separate [PR ](https://github.com/seL4/l4v/pull/852)so that I could do all of the architectures now. Doing that convinced me that I might as well make the generic version and then more specific arch ones as needed.

lsf37 commented on proof/invariant-abstract/AARCH64/Machine_AI.thy:
```diff
@@ -267,16 +271,16 @@ crunch_ignore (valid, empty_fail, no_fail)
    with the following manual interventions:
-   - remove false positives: get_def, gets_def, getFPUState, getRegister, getRestartPC
-   - add read_cntpct
+   - remove false positives: get_def, gets, gets_def, getFPUState, getRegister, getRestartPC
+   - add read_cntpct and readFpuState
```
Thanks for updating that comment!

lsf37 reviewed (approved):
Good work, and a very substantial chunk of the overall FPU proof. Only the minor comments above from me, no real content changes.

corlewis commented:
For anyone reviewing this in its current state, ignore the `refactor arch_thread_get/set lemmas` commit, I've pulled it into a separate PR for master and it will go away here once https://github.com/seL4/l4v/pull/852 is merged and I've rebased this branch.

corlewis commented:
> For anyone reviewing this in its current state, ignore the `refactor arch_thread_get/set lemmas` commit, I've pulled it into a separate PR for master and it will go away here once #852 is merged and I've rebased this branch.

This branch has been rebased over master now #852 has been merged.
I've also updated x64 aspec for the `tcb_cur_fpu` ghost variable, as it will be needed there in the future and I want to keep the architectures in sync where possible to make patching between them easier once I'm up to that stage.

Xaphiosis commented on proof/invariant-abstract/Tcb_AI.thy:
```diff
@@ -1215,6 +1265,13 @@ lemma decode_unbind_notification_wf:
+   decode_set_flags args (cap.ThreadCap t)
+   \<lbrace>tcb_inv_wf\<rbrace>,-"
+  apply (simp add: decode_set_flags_def)
+  by wpsimp
```
trivial: apply+by can be one-lined into by .../..., same as above

Xaphiosis commented on proof/invariant-abstract/Tcb_AI.thy:
```diff
@@ -816,24 +824,64 @@ lemma bind_notification_invs:
+             thread_set_valid_ioc_trivial thread_set_valid_arch_state thread_set_valid_cur_fpu
+             thread_set_cap_refs_respects_device_region
+              | simp add: ran_tcb_cap_cases invs_def valid_state_def valid_pspace_def
+              | rule conjI | erule disjE)+
```
this looks like a move, but the indentation is off on wp args and `|`

Xaphiosis commented on proof/invariant-abstract/Tcb_AI.thy:
```diff
@@ -41,6 +41,12 @@ locale Tcb_AI_1 =
+  assumes arch_prepare_set_domain_invs[wp]:
+    "\<And>t d. arch_prepare_set_domain t d \<lbrace>invs :: 'state_ext state \<Rightarrow> _\<rbrace>"
+  assumes arch_prepare_set_domain_typ_at[wp]:
+    "\<And>t d P T p. arch_prepare_set_domain t d \<lbrace>\<lambda>s::'state_ext state. P (typ_at T p s)\<rbrace>"
```
indent of your new assumptions does not match the rest in the block; might want to blend in

Xaphiosis commented on proof/invariant-abstract/DetSchedSchedule_AI.thy:
```diff
@@ -1552,6 +1575,40 @@ lemma thread_set_not_state_valid_sched:
+
+lemma arch_thread_set_valid_sched_action[wp]:
+  "arch_thread_set f tptr \<lbrace>valid_sched_action\<rbrace>"
+  by (wp valid_sched_action_lift)
```
curious: are any in this block cruncheable?

Xaphiosis commented on proof/invariant-abstract/DetSchedSchedule_AI.thy:
```diff
@@ -220,6 +220,7 @@ abbreviation valid_sched_except_blocked_2 where
 abbreviation valid_sched_except_blocked :: "'z state \<Rightarrow> bool" where
   "valid_sched_except_blocked s \<equiv> valid_sched_except_blocked_2 (ready_queues s) (etcbs_of s) (scheduler_action s) (cur_domain s) (kheap s) (cur_thread s) (idle_thread s)"
 
+\<comment> \<open>FIXME: remove\<close>
```
for this PR?

corlewis replied:
It's not really related to this PR so I wasn't planning to, but it violates enough of our guidelines and was easy enough that I've now done it.

Xaphiosis commented on proof/invariant-abstract/ARM/ArchVSpaceEntries_AI.thy:
```diff
@@ -1521,8 +1521,9 @@ lemma handle_invocation_valid_pdpt[wp]:
-       switch_to_idle_thread, schedule_choose_new_thread
+crunch
+  handle_event, activate_thread, switch_to_thread, switch_to_idle_thread,
+  schedule_choose_new_thread, arch_prepare_next_domain
```
is this better? it'll cause the context block for diffs to show only "crunch"

corlewis replied:
It looks like we don't have any official style guidelines for crunch but this is what I've always done when the constants being crunched are too long to fit on the same line as the crunch command. I don't know how many of them are due to me but there's quite a lot of examples of this style in the proofs.

Thinking through it now, is it useful for the context block in diffs to show the first few constants being crunched, with no indication that there's more beyond a trailing comma?

Xaphiosis replied:
I guess that's fair. It might still be worth it to have something so that you can search for that particular line in a file, but in a pinch the diff markers should tell you where that is. So I guess apart from a thought experiment, I don't have anything better for this one.

Xaphiosis commented on proof/invariant-abstract/ARM/ArchInvariants_AI.thy:
```diff
@@ -747,6 +747,12 @@ definition
   "hyp_live ko \<equiv> False"
 
+definition arch_tcb_live :: "arch_tcb \<Rightarrow> bool" where
+  "arch_tcb_live arch_tcb \<equiv> False"
```
usually we want a comment for these, e.g. "No FPU on this platform" or whatever is the reason

Xaphiosis commented on proof/invariant-abstract/AARCH64/ArchTcbAcc_AI.thy:
```diff
@@ -231,6 +233,34 @@ lemma arch_thread_set_vcpus_of[wp]:
   done
 
+\<comment> \<open>FIXME: Shouldn't state_hyp_refs_of be generic? If it was, then this and
+          arch_thread_set_sym_refs_hyp' could be generic as well\<close>
```
good question. I have no idea, when I did the split it looked like it shouldn't be, or at least its definition shouldn't be. @lsf37 do you remember discussing this by any chance?

Xaphiosis replied:
also `FIXME arch-split` please

Xaphiosis replied:
ok, that sounds promising then

Xaphiosis commented on proof/invariant-abstract/AARCH64/ArchSchedule_AI.thy:
```diff
@@ -28,9 +28,88 @@ lemma dmo_mapM_storeWord_0_invs[wp,Schedule_AI_assms]:
   by (simp add: upto.simps word_bits_def)
 
+lemma set_arm_current_fpu_owner_valid_cur_fpu_except[wp]:
+  "set_arm_current_fpu_owner new_owner \<lbrace>valid_cur_fpu\<rbrace>"
```
name doesn't match statement (except)

Xaphiosis commented on proof/invariant-abstract/AARCH64/ArchKHeap_AI.thy:
```diff
@@ -828,6 +828,34 @@ lemma valid_arch_state_lift_aobj_at:
 end
 
+\<comment> \<open>FIXME: if tcb_at and related terms were moved to InvariantsPre_AI then this lemma could go in
+     ArchInvariants_AI.\<close>
```
you might want to make this and the one below a `FIXME arch-split`, I'm not sure right now

Xaphiosis commented on proof/invariant-abstract/AARCH64/ArchInvariants_AI.thy:
```diff
@@ -535,6 +542,9 @@ definition hyp_live :: "kernel_object \<Rightarrow> bool" where
    |  _ \<Rightarrow> False"
 
+definition arch_tcb_live :: "arch_tcb \<Rightarrow> bool" where
+  "arch_tcb_live arch_tcb \<equiv> tcb_cur_fpu arch_tcb"
```
I think some kind of comment would be useful in this file that indicates the magic link between arch_tcb_live, is_tcb_cur_fpu and valid_cur_fpu, indicating whether a thread is the current FPU owner via ghost state and also indicating which thread is the FPU owner via global state

corlewis replied:
I've added a couple of comments, let me know if they help or if you think it needs more. There's also the comment in the spec that is relevant and that people might come across if they're curious about the ghost variable.
https://github.com/seL4/l4v/blob/2cffea962d05fdbb99f9dfea5a27c73512c3bbbd/spec/abstract/AARCH64/Arch_Structs_A.thy#L370-L372

corlewis replied:
Done. Do you think the comment on `valid_cur_fpu` is helpful or is it repeating the definition and not adding much?

Xaphiosis replied:
Cool, thanks.
It does a little, but it documents the intent. Looking at the definition and seeing it matches the intent gives me good feelings about it, so I'd leave it as is.

Xaphiosis commented on proof/invariant-abstract/AARCH64/ArchBCorres_AI.thy:
```diff
@@ -38,6 +38,11 @@ lemma pt_lookup_from_level_bcorres[wp]:
   "bcorres (pt_lookup_from_level l r b c) (pt_lookup_from_level l r b c)"
   by (induct l arbitrary: r b c rule: vm_level.minus_induct; wpsimp simp: pt_lookup_from_level_simps)
 
+\<comment> \<open>FIXME: move to BCorres_UL\<close>
```
this PR?

Xaphiosis reviewed (approved):
Very nice work capturing this. I only have minor comments and questions.

corlewis commented:
Ok, I think all comments and suggestions have now been handled. The two initial commits and the three labelled squash contain all of this round of changes if someone wants to look at them. In particular, the new commit about removing the wp declaration of `valid_idle_etcb_lift` is completely new content.

lsf37 commented:
Looks good from my side, including the new material.

Xaphiosis commented on proof/invariant-abstract/AARCH64/ArchKHeap_AI.thy:
```diff
@@ -828,6 +828,34 @@ lemma valid_arch_state_lift_aobj_at:
+  by (auto simp: valid_cur_fpu_defs is_tcb)
+
+\<comment> \<open>FIXME arch-split: is_tcb_cur_fpu can't be defined like this because arch_tcb_at isn't defined until
+     Invariants_AI.\<close>
```
any reason for this novelty in indentation (and above)? it's showing up as aligning with "X" in "FIXME"

corlewis replied:
No reason, I've tweaked it now. My best guess is that it was a combination of not paying much attention due to it being in a FIXME comment and vaguely wanting to avoid the long hanging indentation.
